### PR TITLE
Refactor TTL providers

### PR DESCRIPTION
**Pull request by DanielNi**
WHAT: This PR combines the TTL providers into a single class `CombinedTtlProvider.java` and changes the `TenantTtlProvider.java` interface slightly so that it is less complicated.

WHY: Cleans up the TTL provider code a bit and adds some testing.

HOW: 
- Added the `CombinedTtlProvider.java` class
- `TenantTtlProvider.java` now returns `Optional`, which is easier to reason with than exceptions
- Removed `set()` method in the providers since it is not currently used and throws an exception
- Moved the 5x TTL logic from `SafetyTtlProvider.java` to `CassandraModel.java` to make it less confusing
- Made the TTL provider tests a bit more robust
- Removed the Hist TTL configs since they are no longer being used

### CODE REVIEW

**Comment by DanielNi** [comment]
Just realized that this changes the behavior/intent of the providers in a way that might not be ideal, so I'm going to tweak it a bit.

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        long ttl;
+        if (gran == Granularity.FULL && primary.areTTLsForced()) {
+            ttl = primary.getConfigTTLForIngestion().toMillis();
+        } else {
+            ttl = safety.getTTL(tenantid, gran, RollupType.BF_BASIC).get().toMillis();
+        }
+        return ttl;
+    }
+
+    private static Optional<TimeValue> getMaximum(Optional<TimeValue> primaryValue, Optional<TimeValue> safetyValue) {
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java]: do we want to change the name on this?  To me, getMaximum() implies a comparison, which isn't really happening.  getTTL() ?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java [hunk 2]
```diff
@@ -41,98 +43,103 @@ public static ConfigTtlProvider getInstance() {
...
         // String rollups
-        stringTTL = new TimeValue(config.getIntegerProperty(TtlConfig.STRING_METRICS_TTL), TimeUnit.DAYS);
+        TimeValue stringTTL = null;
+        try {
+            int value = config.getIntegerProperty(TtlConfig.STRING_METRICS_TTL);
+            if (value > 0) {
+                stringTTL = new TimeValue(config.getIntegerProperty(TtlConfig.STRING_METRICS_TTL), TimeUnit.DAYS);
+            }
+        } catch (NumberFormatException ignored) {
+        }
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java]: would we want to log this exception?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java [hunk 3]
```diff
@@ -41,98 +43,103 @@ public static ConfigTtlProvider getInstance() {
...
+        Configuration config,
+        Granularity granularity,
+        RollupType rollupType,
+        TtlConfig configKey) {
+        int value;
+        try {
+            value = config.getIntegerProperty(configKey);
+            if (value < 0) return false;
+        } catch (NumberFormatException e) {
+            return false;
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java]: log this?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java [hunk 4]
```diff
@@ -44,30 +44,20 @@
...
+    public static final MetricColumnFamily CF_METRICS_STRING = new MetricColumnFamily(CF_METRICS_STRING_NAME, new TimeValue(365 * 3 * 5, TimeUnit.DAYS));
+
+    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_FULL = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_FULL_NAME, new TimeValue(5, TimeUnit.DAYS));
+    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_5M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_5M_NAME, new TimeValue(10, TimeUnit.DAYS));
+    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_20M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_20M_NAME, new TimeValue(20, TimeUnit.DAYS));
+    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_60M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_60M_NAME, new TimeValue(31 * 5, TimeUnit.DAYS));
+    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_240M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_240M_NAME, new TimeValue(60 * 5, TimeUnit.DAYS));
+    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_1440M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_1440M_NAME, new TimeValue(365 * 5, TimeUnit.DAYS));
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java]: :+1 - i'm not verify familiar with this, but much cleaner.

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java (hunk 1 above)
Comment (DanielNi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java]: I think the original idea was that the safety values would be a lower bound, but I agree that this name is confusing. Besides, I don't think the lower bound is enforced anywhere (probably something worth doing).

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java [hunk 5]
```diff
@@ -41,98 +43,105 @@ public static ConfigTtlProvider getInstance() {
...
         // String rollups
-        stringTTL = new TimeValue(config.getIntegerProperty(TtlConfig.STRING_METRICS_TTL), TimeUnit.DAYS);
+        TimeValue stringTTL = null;
+        try {
+            int value = config.getIntegerProperty(TtlConfig.STRING_METRICS_TTL);
+            if (value > 0) {
+                stringTTL = new TimeValue(config.getIntegerProperty(TtlConfig.STRING_METRICS_TTL), TimeUnit.DAYS);
+            }
+        } catch (NumberFormatException ex) {
+            log.warn("No valid String TTL in config.");
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java]: I'd include the exception as well.  E.g., http://www.slf4j.org/api/org/slf4j/Logger.html#warn(java.lang.String, java.lang.Throwable)

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java [hunk 6]
```diff
@@ -41,98 +43,105 @@ public static ConfigTtlProvider getInstance() {
...
+        Configuration config,
+        Granularity gran,
+        RollupType rollupType,
+        TtlConfig configKey) {
+        int value;
+        try {
+            value = config.getIntegerProperty(configKey);
+            if (value < 0) return false;
+        } catch (NumberFormatException e) {
+            log.info("No valid TTL config set for granularity: {}, rollup type: {}", gran.name(), rollupType.name());
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java]: Please include the exception in the log statement.

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/SafetyTtlProvider.java [hunk 7]
```diff
@@ -27,79 +29,51 @@
...
-                try {
-                    MetricColumnFamily metricCF = CassandraModel.getColumnFamily(RollupType.classOf(type, granularity), granularity);
-                    TimeValue ttl = new TimeValue(metricCF.getDefaultTTL().getValue() * 5, metricCF.getDefaultTTL().getUnit());
-                    ttlMapBuilder.put(granularity, type, ttl);
-                } catch (IllegalArgumentException ex) {
-
+                if (type == RollupType.NOT_A_ROLLUP) {
+                    continue;
                 }
+                MetricColumnFamily metricCF = CassandraModel.getColumnFamily(RollupType.classOf(type, granularity), granularity);
```
Comment (usnavi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/SafetyTtlProvider.java]: @DanielNi  - I missed this before, but FYI, we are currently moving to the Datastax driver, which will allow us to upgrade cassandra.  MetricColumnFamily sub-classes an Astyanax-specific class.  Long-term we'll want to wrap driver-specific types in a class, e.g., https://github.com/rackerlabs/blueflood/blob/master/blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DRawIO.java

However, this will probably happen after we accept this PR, so we can handle this later.

cc - @ChandraAddala @shintasmith @VinnyQ

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        long ttl;
+        if (gran == Granularity.FULL && primary.areTTLsForced()) {
+            ttl = primary.getConfigTTLForIngestion().toMillis();
+        } else {
+            ttl = safety.getTTL(tenantid, gran, RollupType.BF_BASIC).get().toMillis();
+        }
+        return ttl;
+    }
+
+    private static Optional<TimeValue> getTimeValue(Optional<TimeValue> primaryValue, Optional<TimeValue> safetyValue) {
```
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java]: Why does this need to be ```static```?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            ConfigTtlProvider.getInstance(),
+            SafetyTtlProvider.getInstance()
+    );
+
+    public static CombinedTtlProvider getInstance() {
+        return INSTANCE;
+    }
+
+    @VisibleForTesting
+    CombinedTtlProvider(ConfigTtlProvider primary, SafetyTtlProvider safety) {
```
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java]: 👍

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java [hunk 10]
```diff
@@ -41,98 +43,106 @@ public static ConfigTtlProvider getInstance() {
...
+    /**
+     * Helper function to build the ttl mapping. Only insert to the mapping if the value is a valid date.
+     * @param ttlMapBuilder
+     * @param config
+     * @param gran
+     * @param rollupType
+     * @param configKey
+     * @return true if the insertion is successful, false otherwise.
+     */
+    static boolean put(
```
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/ConfigTtlProvider.java]: Does this method needs to be ```static``` ?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java [hunk 11]
```diff
@@ -44,30 +44,20 @@
...
-
-    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_FULL = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_FULL_NAME, new TimeValue(1, TimeUnit.DAYS));
-    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_5M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_5M_NAME, new TimeValue(2, TimeUnit.DAYS));
-    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_20M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_20M_NAME, new TimeValue(4, TimeUnit.DAYS));
-    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_60M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_60M_NAME, new TimeValue(31, TimeUnit.DAYS));
-    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_240M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_240M_NAME, new TimeValue(60, TimeUnit.DAYS));
-    public static final MetricColumnFamily CF_METRICS_PREAGGREGATED_1440M = new MetricColumnFamily(CF_METRICS_PREAGGREGATED_1440M_NAME, new TimeValue(365, TimeUnit.DAYS));
+    public static final MetricColumnFamily CF_METRICS_FULL = new MetricColumnFamily(CF_METRICS_FULL_NAME, new TimeValue(5, TimeUnit.DAYS));
```
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java]: Why was this changed from 1 day to 5 days?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java (hunk 4 above)
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java]: Actually, all of the above have changed values. Can we leave them as is? You can always configure them in config file, yes?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/Configuration.java [hunk 12]
```diff
@@ -17,25 +17,41 @@
...
+     */
+    public static Configuration getTestInstance(Properties defaultProps) {
+        Configuration config = new Configuration(false);
+        for (Map.Entry<Object, Object> entry : defaultProps.entrySet()) {
+            config.defaultProps.put(entry.getKey(), entry.getValue());
+        }
+        return config;
+    }
+
+    private Configuration(boolean shouldInitialize) {
```
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/Configuration.java]: To be honest, I'm not exactly thrilled on the addition of parameter ```shouldInitialize``` to the constructor. It creates a different behavior of the class that skips the ```init()```, for the sake of testing. However, I don't have a better suggestion other than to mock Configuration class for testing. This has been done in some places in blueflood.

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java (hunk 4 above)
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java]: Sorry, I just realized these values were moved from TtlConfig. I guess we never used the default TTLs that were specified here?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/Configuration.java [hunk 13]
```diff
@@ -17,25 +17,41 @@
 package com.rackspacecloud.blueflood.service;
 
 import com.google.common.collect.Lists;
+import com.rackspacecloud.blueflood.cache.ConfigTtlProvider;
```
Comment (shintasmith) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/Configuration.java]: Do you need this import?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java (hunk 4 above)
Comment (DanielNi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/CassandraModel.java]: Yeah, for some reason they were being multiplied by 5 there...

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java (hunk 8 above)
Comment (DanielNi) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/cache/CombinedTtlProvider.java]: It doesn't have to be, but I suppose it's a matter of style/philosophy. I can change it if you prefer.

**Comment by DanielNi** [comment]
@shintasmith @usnavi Does this look okay? If so, can we merge?

**Comment by shintasmith** [comment]
@DanielNi , this looks good. Merging.
